In [1]:
# importing libraries
import numpy as np
import pandas as pd

In [2]:
#Reading CASE-SHILLER Index into a dataframe
df_CS = pd.read_csv(r"C:\Users\S\Downloads\CASESHELLER.csv")

#Changing dtype of date column
df_CS["DATE"] = pd.to_datetime(df_CS["DATE"])

#Resetting Index
df_CS.reset_index(inplace = True)
df_CS.drop(columns = ["index"], inplace = True)

#Creating "Year" and "Month" columns
df_CS["Year"] = pd.DatetimeIndex(df_CS["DATE"]).year
df_CS["Month"] = pd.DatetimeIndex(df_CS["DATE"]).month
print(df_CS.shape)
df_CS.tail()

(252, 4)


DATE  CSUSHPISA  Year  Month
247 2022-08-01    301.638  2022      8
248 2022-09-01    299.443  2022      9
249 2022-10-01    299.017  2022     10
250 2022-11-01    298.232  2022     11
251 2022-12-01    297.189  2022     12

In [3]:
# Reading Unemployment Rate Data into a dataframe
df_unemp = pd.read_csv(r"C:\Users\S\Downloads\UNEMP_RATE.csv", names = ["DATE", "UNRATE"],  skiprows = 1)
print(df_unemp.shape)
df_unemp.tail()

(252, 2)


DATE  UNRATE
247  2022-08-01     3.7
248  2022-09-01     3.5
249  2022-10-01     3.7
250  2022-11-01     3.6
251  2022-12-01     3.5

In [4]:
# Reading Per Capita GDP Data into a dataframe
df_pcgdp = pd.read_csv(r"C:\Users\S\Downloads\PERCAPITA_GDP.csv", names = ["DATE", "Per_Capita_GDP"], skiprows = 1)
print(df_pcgdp.shape)
df_pcgdp.tail()

(84, 2)


DATE  Per_Capita_GDP
79  2021-10-01         60117.0
80  2022-01-01         59836.0
81  2022-04-01         59688.0
82  2022-07-01         60080.0
83  2022-10-01         60376.0

In [5]:
# Reading Interest Rate Data into a dataframe
df_fed_rate = pd.read_csv(r"C:\Users\S\Downloads\INT_RATE.csv",names = ["DATE", "FEDFUNDS"],  skiprows = 1)
print(df_fed_rate.shape)
df_fed_rate.tail()

(252, 2)


DATE  FEDFUNDS
247  2022-08-01      2.33
248  2022-09-01      2.56
249  2022-10-01      3.08
250  2022-11-01      3.78
251  2022-12-01      4.10

In [6]:
#Reading Construction_price into a dataframe
df_cons_price_index = pd.read_csv(r"C:\Users\S\Downloads\CONSTRUCTION_PRICE.csv", names = ["DATE", "Cons_Materials"], skiprows = 1)
print(df_cons_price_index.shape)
df_cons_price_index.tail()

(252, 2)


DATE  Cons_Materials
247  2022-08-01         342.753
248  2022-09-01         336.464
249  2022-10-01         333.796
250  2022-11-01         330.369
251  2022-12-01         326.449

In [7]:
#Reading  Consumer Price Index in a dataframe
df_CPI = pd.read_csv(r"C:\Users\S\Downloads\CPIAUCSL.csv", names = ["DATE", "CPI"], skiprows = 1)
print(df_CPI.shape)
df_CPI.tail()

(252, 2)


DATE      CPI
247  2022-08-01  295.320
248  2022-09-01  296.539
249  2022-10-01  297.987
250  2022-11-01  298.598
251  2022-12-01  298.990

In [8]:
#Reading  Monthly new house supply in a dataframe
df_house = pd.read_csv(r"C:\Users\S\Downloads\MONTHLY_HOUSE_SUPPLY.csv", names = ["DATE", "Houses"], skiprows = 1)
print(df_house.shape)
df_house.tail()

(252, 2)


DATE  Houses
247  2022-08-01     8.7
248  2022-09-01     9.7
249  2022-10-01     9.7
250  2022-11-01     9.4
251  2022-12-01     8.5

In [9]:
#Reading population above 65 in a dataframe
df_oldpop = pd.read_csv(r"C:\Users\S\Downloads\POPULATION_ABOVE_65.csv", names = ["DATE", "OLD_POPULATION"],skiprows=1)
print(df_oldpop.shape)
df_oldpop.tail()

(21, 2)


DATE  OLD_POPULATION
16  2018-01-01       15.397698
17  2019-01-01       15.791801
18  2020-01-01       16.223400
19  2021-01-01       16.678895
20  2022-01-01       17.128121

In [10]:
# Reading population percent in a dataframe
df_pop = pd.read_csv(r"C:\Users\S\Downloads\TOTAL_POPULATION.csv", names = ["DATE", "POPULATION"], skiprows=1)
print(df_pop.shape)
df_pop.tail()

(21, 2)


DATE  POPULATION
16  2018-01-01   326838199
17  2019-01-01   328329953
18  2020-01-01   331511512
19  2021-01-01   332031554
20  2022-01-01   333287557

In [11]:
# Reading working age population in a dataframe
df_working_pop = pd.read_csv(r"C:\Users\S\Downloads\WORKING_AGE_POP.csv", names = ["DATE", "WORKING_AGE_POP"], skiprows = 1)
print(df_working_pop.shape)
df_working_pop.tail()

(252, 2)


DATE  WORKING_AGE_POP
247  2022-08-01     2.073870e+08
248  2022-09-01     2.074717e+08
249  2022-10-01     2.074577e+08
250  2022-11-01     2.075131e+08
251  2022-12-01     2.075091e+08

In [12]:
# Reading property tax in a dataframe
df_property_tax = pd.read_csv(r"C:\Users\S\Downloads\PROPERTY_TAX.csv", names = ["DATE", "PROPERTY_TAX"], skiprows = 1)
print(df_property_tax.shape)
df_property_tax.tail()

(21, 2)


DATE  PROPERTY_TAX
16  2018-01-01        10.044
17  2019-01-01        10.482
18  2020-01-01        10.905
19  2021-01-01        11.464
20  2022-01-01        11.907

In [13]:
# Merging Per Capita GDP (Quarterly data)
df_pcgdp["DATE"] = pd.to_datetime(df_pcgdp["DATE"])
df_merged = pd.merge(df_CS,df_pcgdp, how = "left")
df_merged.head()

DATE  CSUSHPISA  Year  Month  Per_Capita_GDP
0 2002-01-01    117.143  2002      1         46683.0
1 2002-02-01    117.845  2002      2             NaN
2 2002-03-01    118.686  2002      3             NaN
3 2002-04-01    119.610  2002      4         46866.0
4 2002-05-01    120.724  2002      5             NaN

# The data is quarterly. We will impute for other months using linear interpolation after we create the final dataframe combining all the data.

In [14]:
# Concating dataframes having monthly data to create one dataframe
df = pd.DataFrame()
df_bymonth = [df_merged, df_unemp,df_fed_rate,df_cons_price_index,df_CPI,df_working_pop,df_house]
for df1 in df_bymonth:
    df1["DATE"] = pd.to_datetime(df1["DATE"])
    df1 = df1.set_index("DATE")
    df = pd.concat([df,df1], axis = 1)
print(df.shape)
df.tail()

(252, 10)


CSUSHPISA  Year  Month  Per_Capita_GDP  UNRATE  FEDFUNDS   
DATE                                                                   
2022-08-01    301.638  2022      8             NaN     3.7      2.33  \
2022-09-01    299.443  2022      9             NaN     3.5      2.56   
2022-10-01    299.017  2022     10         60376.0     3.7      3.08   
2022-11-01    298.232  2022     11             NaN     3.6      3.78   
2022-12-01    297.189  2022     12             NaN     3.5      4.10   

            Cons_Materials      CPI  WORKING_AGE_POP  Houses  
DATE                                                          
2022-08-01         342.753  295.320     2.073870e+08     8.7  
2022-09-01         336.464  296.539     2.074717e+08     9.7  
2022-10-01         333.796  297.987     2.074577e+08     9.7  
2022-11-01         330.369  298.598     2.075131e+08     9.4  
2022-12-01         326.449  298.990     2.075091e+08     8.5

In [15]:
print(df_oldpop.columns)
print(df_pop.columns)
print(df_property_tax.columns)

Index(['DATE', 'OLD_POPULATION'], dtype='object')
Index(['DATE', 'POPULATION'], dtype='object')
Index(['DATE', 'PROPERTY_TAX'], dtype='object')


In [16]:
# Merging other dataframes
others = [df_pop, df_oldpop, df_property_tax]
for df1 in others:
    if "Year" not in df1.columns:
        df1["Year"] = pd.DatetimeIndex(df1["DATE"]).year
        df1.set_index("DATE", inplace = True)
        df = pd.merge(df, df1, how = "left", on = "Year")
    else:
        df1.set_index("DATE", inplace = True)
        df = pd.merge(df, df1, how = "left", on = "Year")
df["DATE"] = df_CS["DATE"]
df.set_index("DATE", inplace = True)
df.head()



CSUSHPISA  Year  Month  Per_Capita_GDP  UNRATE  FEDFUNDS   
DATE                                                                   
2002-01-01    117.143  2002      1         46683.0     5.7      1.73  \
2002-02-01    117.845  2002      2             NaN     5.7      1.74   
2002-03-01    118.686  2002      3             NaN     5.7      1.73   
2002-04-01    119.610  2002      4         46866.0     5.9      1.75   
2002-05-01    120.724  2002      5             NaN     5.8      1.75   

            Cons_Materials    CPI  WORKING_AGE_POP  Houses  POPULATION   
DATE                                                                     
2002-01-01           142.0  177.7     1.825638e+08     4.2   287625193  \
2002-02-01           142.2  178.0     1.827945e+08     4.0   287625193   
2002-03-01           143.2  178.5     1.830751e+08     4.1   287625193   
2002-04-01           143.5  179.3     1.832577e+08     4.3   287625193   
2002-05-01           143.8  179.5     1.835080e+08     4.0   287625193   

            OLD_POPULATION  PROPERTY_TAX  
DATE                                      
2002-01-01       12.287458         5.168  
2002-02-01       12.287458         5.168  
2002-03-01       12.287458         5.168  
2002-04-01       12.287458         5.168  
2002-05-01       12.287458         5.168

In [17]:
df.shape

(252, 13)

In [18]:
df.isna().sum()

CSUSHPISA            0
Year                 0
Month                0
Per_Capita_GDP     168
UNRATE               0
FEDFUNDS             0
Cons_Materials       0
CPI                  0
WORKING_AGE_POP      0
Houses               0
POPULATION           0
OLD_POPULATION       0
PROPERTY_TAX         0
dtype: int64

# We hae 168 Nan Values in Per_capita_GDP

In [26]:
# Filling missing values in the Per_Capita_GDP column using linear interpolation
df["Per_Capita_GDP"] = df["Per_Capita_GDP"].interpolate()

In [27]:
df.head()

CSUSHPISA  Year  Month  Per_Capita_GDP  UNRATE  FEDFUNDS   
DATE                                                                   
2002-01-01    117.143  2002      1    46683.000000     5.7      1.73  \
2002-02-01    117.845  2002      2    46744.000000     5.7      1.74   
2002-03-01    118.686  2002      3    46805.000000     5.7      1.73   
2002-04-01    119.610  2002      4    46866.000000     5.9      1.75   
2002-05-01    120.724  2002      5    46889.333333     5.8      1.75   

            Cons_Materials    CPI  WORKING_AGE_POP  Houses  POPULATION   
DATE                                                                     
2002-01-01           142.0  177.7     1.825638e+08     4.2   287625193  \
2002-02-01           142.2  178.0     1.827945e+08     4.0   287625193   
2002-03-01           143.2  178.5     1.830751e+08     4.1   287625193   
2002-04-01           143.5  179.3     1.832577e+08     4.3   287625193   
2002-05-01           143.8  179.5     1.835080e+08     4.0   287625193   

            OLD_POPULATION  PROPERTY_TAX  
DATE                                      
2002-01-01       12.287458         5.168  
2002-02-01       12.287458         5.168  
2002-03-01       12.287458         5.168  
2002-04-01       12.287458         5.168  
2002-05-01       12.287458         5.168

In [28]:
df.dropna(inplace=True)

In [29]:
df.isna().sum()

CSUSHPISA          0
Year               0
Month              0
Per_Capita_GDP     0
UNRATE             0
FEDFUNDS           0
Cons_Materials     0
CPI                0
WORKING_AGE_POP    0
Houses             0
POPULATION         0
OLD_POPULATION     0
PROPERTY_TAX       0
dtype: int64

In [24]:
df.shape

(252, 13)

In [25]:
df.tail(10)

CSUSHPISA  Year  Month  Per_Capita_GDP  UNRATE  FEDFUNDS   
DATE                                                                   
2022-03-01    296.271  2022      3    59737.333333     3.6      0.20  \
2022-04-01    300.625  2022      4    59688.000000     3.6      0.33   
2022-05-01    303.904  2022      5    59818.666667     3.6      0.77   
2022-06-01    304.817  2022      6    59949.333333     3.6      1.21   
2022-07-01    304.108  2022      7    60080.000000     3.5      1.68   
2022-08-01    301.638  2022      8    60178.666667     3.7      2.33   
2022-09-01    299.443  2022      9    60277.333333     3.5      2.56   
2022-10-01    299.017  2022     10    60376.000000     3.7      3.08   
2022-11-01    298.232  2022     11    60376.000000     3.6      3.78   
2022-12-01    297.189  2022     12    60376.000000     3.5      4.10   

            Cons_Materials      CPI  WORKING_AGE_POP  Houses  POPULATION   
DATE                                                                       
2022-03-01         345.852  287.472     2.069999e+08     7.0   333287557  \
2022-04-01         343.786  288.611     2.070599e+08     8.5   333287557   
2022-05-01         353.015  291.268     2.072717e+08     8.3   333287557   
2022-06-01         349.800  294.728     2.074014e+08     9.5   333287557   
2022-07-01         346.790  294.628     2.074207e+08    10.1   333287557   
2022-08-01         342.753  295.320     2.073870e+08     8.7   333287557   
2022-09-01         336.464  296.539     2.074717e+08     9.7   333287557   
2022-10-01         333.796  297.987     2.074577e+08     9.7   333287557   
2022-11-01         330.369  298.598     2.075131e+08     9.4   333287557   
2022-12-01         326.449  298.990     2.075091e+08     8.5   333287557   

            OLD_POPULATION  PROPERTY_TAX  
DATE                                      
2022-03-01       17.128121        11.907  
2022-04-01       17.128121        11.907  
2022-05-01       17.128121        11.907  
2022-06-01       17.128121        11.907  
2022-07-01       17.128121        11.907  
2022-08-01       17.128121        11.907  
2022-09-01       17.128121        11.907  
2022-10-01       17.128121        11.907  
2022-11-01       17.128121        11.907  
2022-12-01       17.128121        11.907

# This is our preprocessed datset. Let's save it as "prepared_dataset.csv".

In [30]:
df.to_csv(("prepared_dataset.csv"))

In [31]:
df = pd.read_csv("prepared_dataset.csv").set_index("DATE")
print(df.shape)
df.head(10)


(252, 13)


CSUSHPISA  Year  Month  Per_Capita_GDP  UNRATE  FEDFUNDS   
DATE                                                                   
2002-01-01    117.143  2002      1    46683.000000     5.7      1.73  \
2002-02-01    117.845  2002      2    46744.000000     5.7      1.74   
2002-03-01    118.686  2002      3    46805.000000     5.7      1.73   
2002-04-01    119.610  2002      4    46866.000000     5.9      1.75   
2002-05-01    120.724  2002      5    46889.333333     5.8      1.75   
2002-06-01    121.812  2002      6    46912.666667     5.8      1.75   
2002-07-01    122.887  2002      7    46936.000000     5.8      1.73   
2002-08-01    123.830  2002      8    46918.333333     5.7      1.74   
2002-09-01    124.779  2002      9    46900.666667     5.7      1.75   
2002-10-01    125.734  2002     10    46883.000000     5.7      1.75   

            Cons_Materials    CPI  WORKING_AGE_POP  Houses  POPULATION   
DATE                                                                     
2002-01-01           142.0  177.7     1.825638e+08     4.2   287625193  \
2002-02-01           142.2  178.0     1.827945e+08     4.0   287625193   
2002-03-01           143.2  178.5     1.830751e+08     4.1   287625193   
2002-04-01           143.5  179.3     1.832577e+08     4.3   287625193   
2002-05-01           143.8  179.5     1.835080e+08     4.0   287625193   
2002-06-01           144.3  179.6     1.836709e+08     4.2   287625193   
2002-07-01           144.7  180.0     1.838901e+08     4.2   287625193   
2002-08-01           145.1  180.5     1.841637e+08     4.0   287625193   
2002-09-01           145.4  180.8     1.844248e+08     3.9   287625193   
2002-10-01           145.0  181.2     1.845547e+08     4.0   287625193   

            OLD_POPULATION  PROPERTY_TAX  
DATE                                      
2002-01-01       12.287458         5.168  
2002-02-01       12.287458         5.168  
2002-03-01       12.287458         5.168  
2002-04-01       12.287458         5.168  
2002-05-01       12.287458         5.168  
2002-06-01       12.287458         5.168  
2002-07-01       12.287458         5.168  
2002-08-01       12.287458         5.168  
2002-09-01       12.287458         5.168  
2002-10-01       12.287458         5.168